### PR TITLE
[drape] Fixed MutableLabelHandle mutation buffer's size.

### LINKED ISSUE #8108: [android] [GWP-ASan]: Buffer Overflow dp::vulkan::VulkanGPUBuffer::UpdateData()
Brand:Google
Model:Pixel 4a
Orientation:Unknown
RAM free: 70.65 MB
Disk free: 998.14 MB
Operating system
Version:Android 13
Orientation:Unknown
Rooted:No
Date:2 May 2024, 14:08:03
App version:2024.04.29-31-Google-beta (24042931)

```
[GWP-ASan]: Buffer Overflow, 0 bytes right of a 256-byte allocation at 0x73f0c0cf00: SIGSEGV  0x00000073f0c0d000
#00 pc 0x4d7dc libc.so (BuildId: 4e07915368c859b1910c68c84a8de75f)
#01 pc 0x76b2a4 liborganicmaps.so (dp::vulkan::VulkanGPUBuffer::UpdateData(void*, void const*, unsigned int, unsigned int) [string.h:61]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#02 pc 0x751f54 liborganicmaps.so (dp::VertexArrayBuffer::ApplyMutation(ref_ptr<dp::GraphicsContext>, ref_ptr<dp::IndexBufferMutator>, ref_ptr<dp::AttributeBufferMutator>) [data_buffer.cpp:84]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#03 pc 0x6eeb2c liborganicmaps.so (std::__ndk1::__function::__func<gui::ShapeRenderer::Render(ref_ptr<dp::GraphicsContext>, ref_ptr<gpu::ProgramManager>, ScreenBase const&)::$_0, std::__ndk1::allocator<gui::ShapeRenderer::Render(ref_ptr<dp::GraphicsContext>, ref_ptr<gpu::ProgramManager>, ScreenBase const&)::$_0>, void (gui::ShapeControl::ShapeInfo&)>::operator()(gui::ShapeControl::ShapeInfo&) [shape.cpp:120]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#04 pc 0x6ef0d8 liborganicmaps.so (std::__ndk1::__function::__func<gui::ShapeRenderer::ForEachShapeInfo(std::__ndk1::function<void (gui::ShapeControl::ShapeInfo&)> const&)::$_0, std::__ndk1::allocator<gui::ShapeRenderer::ForEachShapeInfo(std::__ndk1::function<void (gui::ShapeControl::ShapeInfo&)> const&)::$_0>, void (gui::ShapeControl&)>::operator()(gui::ShapeControl&) [function.h:510]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#05 pc 0x6cf4e8 liborganicmaps.so (gui::ShapeRenderer::ForEachShapeControl(std::__ndk1::function<void (gui::ShapeControl&)> const&) [function.h:510]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#06 pc 0x6cca5c liborganicmaps.so (gui::ShapeRenderer::Render(ref_ptr<dp::GraphicsContext>, ref_ptr<gpu::ProgramManager>, ScreenBase const&) [shape.cpp:154]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#07 pc 0x6c1890 liborganicmaps.so (gui::LayerRenderer::Render(ref_ptr<dp::GraphicsContext>, ref_ptr<gpu::ProgramManager>, bool, ScreenBase const&) [layer_render.cpp:56]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#08 pc 0x6c0048 liborganicmaps.so (df::FrontendRenderer::RenderScene(ScreenBase const&, bool) [frontend_renderer.cpp:1524]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#09 pc 0x6c243c liborganicmaps.so (df::FrontendRenderer::RenderFrame() [frontend_renderer.cpp:1765]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#10 pc 0x6c5868 liborganicmaps.so (df::FrontendRenderer::Routine::Do() [base_renderer.cpp:64]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#11 pc 0xad93a8 liborganicmaps.so (threads::(anonymous namespace)::RunRoutine(std::__ndk1::shared_ptr<threads::IRoutine>) [thread.cpp:26]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#12 pc 0xae14c8 liborganicmaps.so (void* std::__ndk1::__thread_proxy[abi:v170000]<std::__ndk1::tuple<std::__ndk1::unique_ptr<std::__ndk1::__thread_struct, std::__ndk1::default_delete<std::__ndk1::__thread_struct> >, void (*)(std::__ndk1::shared_ptr<threads::IRoutine>), std::__ndk1::shared_ptr<threads::IRoutine> > >(void*) [invoke.h:394]) (BuildId: 330ca0b7a2e78c7ee280e38487282957f3ca0c2f)
#13 pc 0xb63b0 libc.so (BuildId: 4e07915368c859b1910c68c84a8de75f)
#14 pc 0x530b8 libc.so (BuildId: 4e07915368c859b1910c68c84a8de75f)
```

Not too many details. It was likely my Pixel 4a. I don't know how to reproduce this issue.

### LINKED ISSUE #8127: [ios] Crash on startup
A long tap on the icon + Bookmarks from the menu helped to workaround this crash.

DSYM files to decode ips crash logs are attached as [action artifacts](https://github.com/organicmaps/organicmaps/actions/runs/8952369390)

[Macsymbolicator](https://github.com/inket/MacSymbolicator) helps to symbolicate crash logs easily.

```
Incident Identifier: A83084C7-CD6F-454C-A879-BB9FFEA44A28
CrashReporter Key:   9a9432e42dbaad0410fd276e708e5abab8e0644e
Hardware Model:      iPhone14,2
Process:             Organic Maps [24805]
Path:                /private/var/containers/Bundle/Application/4ED447DD-F5E3-4B1E-B81D-FC10B37C8497/Organic Maps.app/Organic Maps
Identifier:          app.organicmaps
Version:             2024.05.04 (10)
AppStoreTools:       15E204
AppVariant:          1:iPhone14,2:15
Code Type:           ARM-64 (Native)
Role:                Foreground
Parent Process:      launchd [1]
Coalition:           app.organicmaps [2082]

Date/Time:           2024-05-10 00:06:21.3635 +0700
Launch Time:         2024-05-10 00:06:20.9207 +0700
OS Version:          iPhone OS 17.4.1 (21E236)
Release Type:        User
Baseband Version:    3.50.04
Report Version:      104

Exception Type:  EXC_BAD_ACCESS (SIGSEGV)
Exception Subtype: KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Codes: 0x0000000000000001, 0x0000000000000000
VM Region Info: 0 is not in any region.  Bytes before following region: 4307058688
      REGION TYPE                 START - END      [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                   100b88000-100c78000 [  960K] r-x/r-x SM=COW  /var/containers/Bundle/Application/4ED447DD-F5E3-4B1E-B81D-FC10B37C8497/Organic Maps.app/Organic Maps
Termination Reason: SIGNAL 11 Segmentation fault: 11
Terminating Process: exc handler [24805]

Triggered by Thread:  22

...

Thread 22 Crashed:
0   libsystem_platform.dylib      	       0x1edce2af0 _platform_memmove + 144
1   Organic Maps                  	       0x100e7e7b0 dp::VertexArrayBuffer::ApplyMutation(ref_ptr<dp::GraphicsContext>, ref_ptr<dp::IndexBufferMutator>, ref_ptr<dp::AttributeBufferMutator>) (in Organic Maps) (vertex_array_buffer.cpp:398) + 3106736
2   Organic Maps                  	       0x100eb64d0 std::__1::__function::__func<gui::ShapeRenderer::Render(ref_ptr<dp::GraphicsContext>, ref_ptr<gpu::ProgramManager>, ScreenBase const&)::$_2, std::__1::allocator<gui::ShapeRenderer::Render(ref_ptr<dp::GraphicsContext>, ref_ptr<gpu::ProgramManager>, ScreenBase const&)::$_2>, void (gui::ShapeControl::ShapeInfo&)>::operator()(gui::ShapeControl::ShapeInfo&) (in Organic Maps) (function.h:364) + 3335376
3   Organic Maps                  	       0x100eb5b50 std::__1::__function::__func<gui::ShapeRenderer::ForEachShapeInfo(std::__1::function<void (gui::ShapeControl::ShapeInfo&)> const&)::$_3, std::__1::allocator<gui::ShapeRenderer::ForEachShapeInfo(std::__1::function<void (gui::ShapeControl::ShapeInfo&)> const&)::$_3>, void (gui::ShapeControl&)>::operator()(gui::ShapeControl&) (in Organic Maps) (function.h:364) + 3332944
4   Organic Maps                  	       0x100eb593c gui::ShapeRenderer::ForEachShapeControl(std::__1::function<void (gui::ShapeControl&)> const&) (in Organic Maps) (shape.cpp:149) + 3332412
5   Organic Maps                  	       0x100eb6214 gui::ShapeRenderer::Render(ref_ptr<dp::GraphicsContext>, ref_ptr<gpu::ProgramManager>, ScreenBase const&) (in Organic Maps) (shape.cpp:99) + 3334676
6   Organic Maps                  	       0x100f4b1f4 df::FrontendRenderer::RenderScene(ScreenBase const&, bool) (in Organic Maps) (frontend_renderer.cpp:1524) + 3944948
7   Organic Maps                  	       0x100f3c100 df::FrontendRenderer::RenderFrame() (in Organic Maps) (frontend_renderer.cpp:1767) + 3883264
8   Organic Maps                  	       0x100e92700 std::__1::__function::__func<df::BaseRenderer::IterateRenderLoop()::$_0, std::__1::allocator<df::BaseRenderer::IterateRenderLoop()::$_0>, void ()>::operator()() (in Organic Maps) (function.h:364) + 3188480
9   Organic Maps                  	       0x100f4fa98 df::FrontendRenderer::Routine::Do() (in Organic Maps) (frontend_renderer.cpp:2435) + 3963544
10  Organic Maps                  	       0x100e0b0f0 void* std::__1::__thread_proxy[abi:ue170006]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (*)(std::__1::shared_ptr<threads::IRoutine>), std::__1::shared_ptr<threads::IRoutine>>>(void*) (in Organic Maps) (thread.h:238) + 2633968
11  libsystem_pthread.dylib       	       0x1edd99a90 _pthread_start + 136
12  libsystem_pthread.dylib       	       0x1edd98fcc thread_start + 8
```

### PR DESCRIPTION
**Pull request by vng**
Should fix https://github.com/organicmaps/organicmaps/issues/8128 and linked issues.
Fixes #8127
Fixes #8108

### CODE REVIEW

**Review by biodranik — COMMENTED** [review]
The fix doesn't work, the label is not cleared properly
telegram-cloud-photo-size-2-5327980903365990279-y

**Comment by biodranik** [comment]
> Please mention the original change that has caused the regression.

The original change was done 10 years ago.

**Comment by biodranik** [comment]
@vng looks like all max 50 chars/vertexes should always be rendered for mutable labels, including the original string and remaining empty characters. Maybe add a suffix from a space char for up to max 50 chars limit?

**Comment by biodranik** [comment]
> Why, then, did this issue emerge in this release?

It has been reproduced in many other releases since the beginning of OM. Now we have got a detailed stack trace and a better understanding of what's going on.

File: drape_frontend/gui/gui_text.cpp [hunk 1]
```diff
@@ -473,17 +484,18 @@ void MutableLabelHandle::GetAttributeMutation(ref_ptr<dp::AttributeBufferMutator
   m_size.y = result.m_boundRect.SizeY();
 
   size_t const byteCount = result.m_buffer.size() * sizeof(MutableLabel::DynamicVertex);
-
   auto const dataPointer = static_cast<MutableLabel::DynamicVertex *>(mutator->AllocateMutationBuffer(byteCount));
   std::copy(result.m_buffer.begin(), result.m_buffer.end(), dataPointer);
 
   dp::BindingInfo const & binding = MutableLabel::DynamicVertex::GetBindingInfo();
-  dp::OverlayHandle::TOffsetNode offsetNode = GetOffsetNode(binding.GetID());
+  dp::OverlayHandle::TOffsetNode const node = GetOffsetNode(binding.GetID());
```
Comment (biodranik) [inline drape_frontend/gui/gui_text.cpp:491]:
```suggestion
  dp::OverlayHandle::TOffsetNode const & node = GetOffsetNode(binding.GetID());
```

File: drape_frontend/gui/gui_text.cpp [hunk 2]
```diff
@@ -374,10 +374,21 @@ void MutableLabel::Precache(PrecacheParams const & params, PrecacheResult & resu
...
+    /// @todo I don't see a better way to clear cached verticies from previous frame.
+    text.append(m_maxLength - sz, ' ');
+  }
+  else if (sz > m_maxLength)
+  {
+    text.erase(m_maxLength - 3);
+    text.append("...");
+  }
+
+  strings::UniString uniText = bidi::log2vis(text);
```
Comment (biodranik) [inline drape_frontend/gui/gui_text.cpp:391]:
```suggestion
  strings::UniString const uniText = bidi::log2vis(text);
```

**Review by biodranik — APPROVED** [review]
I wonder how it was cleared before? :)

**Comment by vng** [comment]
> I wonder how it was cleared before? :)

IDK, the mutator always read m_maxLength from the buffer, luckily to have some stuff in this buffer, that was interpreted as empty vertices ..

**Review by biodranik — APPROVED** [review]

**Pull request merged**